File: 二分查找binarySearch/11. Container With Most Water.java
package 二分查找binarySearch;

/**
 * 求最大面積
 * 
 * Input: height = [1,8,6,2,5,4,8,3,7]
 * Output: 49
 * Explanation: The above vertical lines are represented by array
 * [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the
 * container can contain is 49.
 * 
 * Input: height = [1,1]
 * Output: 1
 * 
 * 理解 
 * 用二分法搜尋
 * 尋找兩個最小的高
 * 乘上他們之間的index
 * 就可以算上他們的面積
 * 在用 max 統計最大值
 */

class Solution {
    public int maxArea(int[] height) {
        // Maximum area will be stored in this variable
        int maximumArea = Integer.MIN_VALUE;
        // Left and right pointers
        int left = 0;
        int right = height.length - 1;
        // Loop until left and right meet
        while (left < right) {
            // Shorter pole/vertical line
            int shorterLine = Math.min(height[left], height[right]);
            // Update maximum area if required
            maximumArea = Math.max(maximumArea, shorterLine * (right - left));
            // If there is a longer vertical line present
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return maximumArea;
    }

}

File: TowPoint雙指針/11. Container With Most Water.kt


解决这个问题的关键在于使用双指针法。
开始时，我们将左指针指向数组的开头，将右指针指向数组的末尾，表示当前容器的边界。
然后，我们计算当前容器的面积，并将其与已知的最大面积进行比较。

接着，我们将较短的线段向内移动一个位置，这是因为如果移动较长的线段，容器的宽度会减小，
因此容纳的水量不会增加。而移动较短的线段可能会找到更高的线段，从而有可能增加容器的高度。

我们一直重复这个过程，直到左指针大于等于右指针，表示无法再找到更高的线段，计算过程结束。

这个算法的时间复杂度是 O(n)，其中 n 是输入数组的长度。
希望这个解释和示例代码有助于理解 "11. Container With Most Water" 问题的解决方法。

class Solution {
    fun maxArea(height: IntArray): Int {
        var left = 0 // 左边界指针
        var right = height.size - 1 // 右边界指针
        var maxArea = 0 // 最大容器面积
        
        while (left < right) {
            val hLeft = height[left]
            val hRight = height[right]
            val h = minOf(hLeft, hRight) // 选取较短的高度
            val w = right - left // 宽度
            
            val area = h * w // 计算当前容器的面积
            maxArea = maxOf(maxArea, area) // 更新最大面积
            
            // 移动较短的指针，以期望找到更高的线段
            if (hLeft < hRight) {
                left++
            } else {
                right--
            }
        }
        return maxArea
    }
}
